### PR TITLE
Reading layer files with no parallelization

### PR DESCRIPTION
**Pull request by GalRogozinski**
Reading a file from multiple threads is expected to reduce performance

Needs to be tested

### CODE REVIEW

**Comment by alon-e** [comment]
>Reading a file from multiple threads is expected to reduce performance

to clarify this is reading multiple (different) files from multiple threads.
would performance be hindered also in the case of very large layer files?
I'll be testing it now.

File: compass/milestone/MilestoneDatabase.java [hunk 1]
```diff
@@ -75,18 +76,46 @@ public MilestoneDatabase(SpongeFactory.Mode powMode, SignatureSource signatureSo
...
+      if (batch.size() < batchSize) {
+        return result;
+      }
+    }
+    return result;
+  }
+
+  private static List<String> readLinesInBatches(BufferedReader br, int batchSize) throws IOException {
+    List<String> result = new ArrayList<>(batchSize);
+    for (int i = 1; i < batchSize; i++) {
```
Comment (karimodm) [inline compass/milestone/MilestoneDatabase.java]: Is the break condition of the `for` loop correct? Take this edge case: if `batchSize` is 1 no line gets read. Also, if `i` represents the amount of line read, why not starting it at `0`?

File: compass/milestone/MilestoneDatabase.java [hunk 2]
```diff
@@ -75,18 +76,46 @@ public MilestoneDatabase(SpongeFactory.Mode powMode, SignatureSource signatureSo
     this.powMode = powMode;
   }
 
+  private static List<String> readInBatches(BufferedReader br, int batchSize, int totalSize) throws IOException {
+    List<String> result = new ArrayList<>(totalSize);
+    for (int i = 1; i < totalSize + batchSize; i += batchSize) {
```
Comment (karimodm) [inline compass/milestone/MilestoneDatabase.java]: Shouldn't the break condition be "read up to totalSize lines" ? See comment below.

File: compass/milestone/MilestoneDatabase.java [hunk 3]
```diff
@@ -76,18 +76,46 @@ public MilestoneDatabase(SpongeFactory.Mode powMode, SignatureSource signatureSo
...
+      List<String> batch = readLinesInBatches(br, batchSize);
+      result.addAll(batch);
+      if (batch.size() < batchSize) {
+        return result;
+      }
+    }
+    return result;
+  }
+
+  private static List<String> readLinesInBatches(BufferedReader br, int batchSize) throws IOException {
```
Comment (GalRogozinski) [inline compass/milestone/MilestoneDatabase.java]: This method does not reduce memory consumption.
It only distracted me from the real solution on line 80...

File: compass/milestone/MilestoneDatabase.java [hunk 4]
```diff
@@ -76,18 +76,46 @@ public MilestoneDatabase(SpongeFactory.Mode powMode, SignatureSource signatureSo
...
-        });
+    Map<Integer, List<String>> result = new HashMap<>();
+    int batchSize = 1024;
+
+    for (Path p : Files.newDirectoryStream(Paths.get(path))) {
+      int idx = Integer.parseInt(p.toString().split("\\.")[1]);
+      int totalSize = 1 << idx;
+      try {
+        BufferedReader br = new BufferedReader(new FileReader(p.toString()));
+        result.put(idx, readInBatches(br, batchSize, totalSize));
```
Comment (GalRogozinski) [inline compass/milestone/MilestoneDatabase.java]: As discussed,
We now can just read directly to the list without the map

**Pull request merged**